### PR TITLE
making explicit the copy constructor of the Variable

### LINKED ISSUE #3789: [core] Possible memory problem when making copies of Kratos Variables
Concerning the problems detected by @AlejandroCornejo and me in #3647, I have seen that those problems were solved by storing the names of the variables instead of copies of the variables themselves.
See, for example, the commit by @loumalouomega : https://github.com/KratosMultiphysics/Kratos/commit/472b1358756f8924e2cc3e3508d143f31331e0c0#diff-7e199cf7365e9a7ce5f2d422f15ae66f 
If the copy of variables is a problem, we should be aware of it, and maybe we should restrict that use somehow.

### PR DESCRIPTION
**Pull request by RiccardoRossi**
as the title says this attempts to avoid copies of variables, which may give memory errors at destruction in some specific cases.

@KratosMultiphysics/technical-committee 

closes #3789

### CODE REVIEW

File: applications/MeshingApplication/custom_processes/metrics_levelset_process.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@ namespace Kratos
 template<SizeType TDim>
 ComputeLevelSetSolMetricProcess<TDim>::ComputeLevelSetSolMetricProcess(
         ModelPart& rThisModelPart,
-        const Variable<array_1d<double,3>> rVariableGradient,
+        const Variable<array_1d<double,3>>& rVariableGradient,
```
Comment (loumalouomega) [inline applications/MeshingApplication/custom_processes/metrics_levelset_process.cpp]: I thought I fixed this long time ago

**Review by loumalouomega — DISMISSED** [review]

**Comment by philbucher** [comment]
@armingeiser @dbaumgaertner can you please take a look at this branch and also do the necessary changes in the ShapeOptApp?
Thx

**Comment by RiccardoRossi** [comment]
@roigcarlo can u add this to the warnings in the CMAkeLists.txt?

#6339

**Comment by roigcarlo** [comment]
https://github.com/KratosMultiphysics/Kratos/pull/6340

**Comment by philbucher** [comment]
for me the core does not compile
~~~
/home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp: In function ‘void Kratos::ConstraintUtilities::PreComputeExplicitConstraintConstribution(Kratos::ModelPart&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’:
/home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp:157:215: error: no matching function for call to ‘std::unordered_map<long unsigned int, Kratos::Variable<double> >::insert(std::pair<long unsigned int, Kratos::Variable<double> >)’
~~~
@loumalouomega this is your code, can you take a look please?

**Comment by philbucher** [comment]
@sunethwarna the RansApp is also not compiling with this, can you please take a look?

**Comment by philbucher** [comment]
@AFranci same for the PfemFluidApp, can you please take a look?

**Comment by philbucher** [comment]
better but still an error:
~~~
/home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp: In function ‘void Kratos::ConstraintUtilities::PreComputeExplicitConstraintMassAndInertia(Kratos::ModelPart&, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string)’:
/home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp:288:113: error: no matching function for call to ‘std::unordered_map<long unsigned int, Kratos::Variable<double> >::insert(std::pair<long unsigned int, Kratos::Variable<double> >)’
     displacement_variable_map.insert(std::pair<std::size_t, Variable<double>>(r_check_dof_x.Key(), r_mass_dof_x));
~~~

**Comment by loumalouomega** [comment]
> 
> 
> better but still an error:
> 
> ```
> /home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp: In function ‘void Kratos::ConstraintUtilities::PreComputeExplicitConstraintMassAndInertia(Kratos::ModelPart&, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string)’:
> /home/philippb/software/Kratos_prod/kratos/utilities/constraint_utilities.cpp:288:113: error: no matching function for call to ‘std::unordered_map<long unsigned int, Kratos::Variable<double> >::insert(std::pair<long unsigned int, Kratos::Variable<double> >)’
>      displacement_variable_map.insert(std::pair<std::size_t, Variable<double>>(r_check_dof_x.Key(), r_mass_dof_x));
> ```

Ups, there is another method

imagen

**Comment by philbucher** [comment]
now core compiles thx @loumalouomega

**Comment by philbucher** [comment]
@armingeiser I fixed the shapeoptapp
BTW I think the loop I changed could be optimized (different PR), you are accessing the variable every loop-iteration!

**Comment by ddiezrod** [comment]
> @roigcarlo can u add this to the warnings in the CMAkeLists.txt?
> 
> #6339

@RiccardoRossi Just a question, why is this API breaker? Isn't this something purely internal?

**Comment by philbucher** [comment]
> > @roigcarlo can u add this to the warnings in the CMAkeLists.txt?
> > #6339
> 
> Just a question, why is this API breaker? Isn't this something purely internal?

is is an api-breaker if you make copies of variables. This will no longer compile if this is merged
Note that you can already fix it now, then you will not be affected

**Comment by armingeiser** [comment]
Thanks @philbucher - you are right about the loop, will fix it ina separate PR

**Comment by ddiezrod** [comment]
> > > @roigcarlo can u add this to the warnings in the CMAkeLists.txt?
> > > #6339
> > 
> > 
> > Just a question, why is this API breaker? Isn't this something purely internal?
> 
> is is an api-breaker if you make copies of variables. This will no longer compile if this is merged
> Note that you can already fix it now, then you will not be affected

Oh, I see, I thought API-breaker was something more related to IO. Thanks for clarifying.

**Comment by RiccardoRossi** [comment]
The point is that we are experiencing from time to time "random" problems with the deallocation of variables. 
One of the possible problematic sources are copies of variables, hence the reason of being of this PR is to make it more difficult (although not impossible) to create copies. (essentially one should always use "const Variable<DataType>&" in the code to be correct

**Comment by loumalouomega** [comment]
> 
> 
> The point is that we are experiencing from time to time "random" problems with the deallocation of variables.
> One of the possible problematic sources are copies of variables, hence the reason of being of this PR is to make it more difficult (although not impossible) to create copies. (essentially one should always use "const Variable&" in the code to be correct

Nobody can stop the power of the const_cast. The mighty const_cast and the almighty reinterpret_cast

**Comment by philbucher** [comment]
I created #6352 where I port the corrections for the usage of `Variable`

No matter the outcome of this PR I think those changes are for the better, and it will also make the review of this PR easier (will come done to very few changes after corrections are merged)

**Comment by philbucher** [comment]
ping @AFranci

**Comment by AFranci** [comment]
Please @philbucher take a look if these are the changes you were looking for.

**Comment by philbucher** [comment]
Looks good @AFranci thx

@RiccardoRossi if you merge master it should work now

**Comment by roigcarlo** [comment]
Dear @KratosMultiphysics/all 

Since this change is critical to solve problems with variables @KratosMultiphysics/technical-committee aproves this and it will me merged on the 02/03/2020.

**Review by roigcarlo — APPROVED** [review]
Due to merge. 👍

**Pull request merged**